### DISCUSSION TITLE
How to populate flow/platforms/myplatform/fill.json

### DISCUSSION THREAD

**Discussion by gkamendje** [The-OpenROAD-Project/OpenROAD-flow-scripts#1581] [Q&A] [answered]
Assuming that I have a DRC rule that says: Minimum density of Mx area is y%.  How should I populate the fill.json? In particular it is not clear  how I should set the following fields. 
```
               "width":                [2.00, 1.00, 0.58, 0.3],
                "height":               [2.00, 1.00, 0.58, 0.3],
                "space_to_fill":        0.3,
                "space_to_non_fill":    3

```
Are the `width` and `height` entries always taken as a pair? 

The doc says 
```
  "space_to_fill": "<real: spacing between fills in microns>",
   "space_to_non_fill": "<real: spacing to non-fill shapes in microns>"
```
Should the `space_to_fill` be at least the minimum spacing distance for the given layer? 
What are `non-fill` shapes in this context? 

For example with the following entries in the `fill.json` file,
```  
   "METAL1" : {
            "layer":                   61,
            "name":                    "METAL1",
            "dir":                     "H",
            "datatype":                 0,
            "space_to_outline":         70,
            "non-opc": {
                "datatype":             0,
                "width":                [2.00, 1.00, 0.58, 0.3],
                "height":               [2.00, 1.00, 0.58, 0.3],
                "space_to_fill":        0.3,
                "space_to_non_fill":    3
            }
        },
        
        "METAL2" : {
            "layer":                   62,
            "name":                    "METAL2",
            "dir":                     "V",
            "datatype":                 0,
            "space_to_outline":         70,
            "non-opc": {
                "datatype":             0,
                "width":                [2.00, 1.00, 0.58, 0.3],
                "height":               [2.00, 1.00, 0.58, 0.3],
                "space_to_fill":        0.3,
                "space_to_non_fill":    3
            }
        },
```
I get the following error while running Klayout in step 6 to create the merged GDS file 
```
INFO: Reading config file: /path_to_my_platform/fill.json
ERROR: Unrecognized fill:     - LAYER METAL2 RECT ( 3420 4480 ) ( 4420 5480 ) ;
```
What does this error means?

**Comment by maliberty** [comment, marked as answer]
Are the width and height entries always taken as a pair?  Yes
space_to_fill is the spacing between two fill objects
space_to_non_fill is the spacing between a fill object and any other sort of object (eg routing, pins)

The error sounds similar to the one in https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1571.  Perhaps the fix there will solve your problem as well once it is merged.

**Reply by gkamendje** [reply]
Indeed applying the following change to `util/def2stream.py` fixed the issue. 
```
diff --git a/flow/util/def2stream.py b/flow/util/def2stream.py
index f13fbd60..74809759 100644
--- a/flow/util/def2stream.py
+++ b/flow/util/def2stream.py
@@ -55,12 +55,25 @@ rect_pat = re.compile(r'''
   (?P<opc>                      # OPC, None if absent
   \s+\+\ OPC
   )?
-  \s+RECT\
-   \(\ (?P<xlo>\d+)\ (?P<ylo>\d+)\ \)\   # rect lower-left pt
-  \(\ (?P<xhi>\d+)\ (?P<yhi>\d+)\ \)\ ; # rect upper-right pt
+  \s+RECT\s+
+   \(\s+(?P<xlo>\d+)\s+(?P<ylo>\d+)\s+\)\s+   # rect lower-left pt
+   \(\s+(?P<xhi>\d+)\s+(?P<yhi>\d+)\s+\)\s+ ; # rect upper-right pt
   ''',
                       re.VERBOSE)
-

```

**Comment by maliberty** [comment]
Note that we fill as much as possible and not to a particular target percentage.

**Reply by gkamendje** [reply]
After running with the USE_FILL, I still see density violation. What would be the suggested strategy to fix density issue?

**Reply by maliberty** [reply]
Can you describe the problem further or include some screen shots of the affected area?

**Reply by gkamendje** [reply]
The problem could well be related to incomplete DRC rules since my `.lydrc` file is still WIP. Having said that, It can be seen on the attached screen shot that KLayout reports that the M1 min density rule is not satisfied in the `polygon: (12.325,55.08;12.325,75.08;32.325,75.08;32.325,55.08)`
![Fill_Density](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/56942214/6c3bd218-dfde-40ab-a5b9-7a977f979aa9)

**Reply by maliberty** [reply]
Please show just M1 as its hard to see what's happening

**Reply by gkamendje** [reply]
Here with M1 only. Please let me know if that helps.
![Fill_Density_M1_only](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/56942214/1f0384f0-b4d6-4f04-94b4-555fdf2f33b2)

**Reply by maliberty** [reply]
I don't see any fill shapes at all.  Do you see M1 fill shapes in the final DEF?  If so you probably have a layer mapping problem.

**Reply by gkamendje** [reply]
Here is the output log of the density fill Tcl script. IT can be seen that for some reason no filling is added on M1 (no idea why). There seems to be filling added on M2 and M3 However, I cannot identify the fill shapes in the final DEF file. Do these follow a specific naming convention?   
```
OpenROAD v2.0-10646-g0dff30da2 
This program is licensed under the BSD-3 license. See the LICENSE file for details.
Components of this program may be licensed under more restrictive licenses which must be honored.
[WARNING FIN-0010] Skipping layer NW.
[WARNING FIN-0010] Skipping layer SN.
[WARNING FIN-0010] Skipping layer POLY1.
[WARNING FIN-0010] Skipping layer CONT.
[INFO FIN-0003] Filling layer METAL1.
[INFO FIN-0009] Filling 0 areas with non-OPC fill.
[INFO FIN-0004] Total fills: 0.
[WARNING FIN-0010] Skipping layer VIA12.
[INFO FIN-0003] Filling layer METAL2.
[INFO FIN-0009] Filling 11 areas with non-OPC fill.
[INFO FIN-0004] Total fills: 94.
[WARNING FIN-0010] Skipping layer VIA23.
[INFO FIN-0003] Filling layer METAL3.
[INFO FIN-0009] Filling 6 areas with non-OPC fill.
[INFO FIN-0004] Total fills: 161.
[WARNING FIN-0010] Skipping layer VIA34.
[INFO FIN-0003] Filling layer METAL4.
[INFO FIN-0009] Filling 2 areas with non-OPC fill.
[INFO FIN-0004] Total fills: 287.
[WARNING FIN-0010] Skipping layer VIA45.
[INFO FIN-0003] Filling layer METAL5.
[INFO FIN-0009] Filling 1 areas with non-OPC fill.
[INFO FIN-0004] Total fills: 1307.
[WARNING FIN-0010] Skipping layer OVERLAP.
Elapsed time: 0:05.89[h:]min:sec. CPU time: user 5.75 sys 0.13 (99%). Peak memory: 276764KB.
```

**Reply by maliberty** [reply]
In DEF there is a FILLS section

**Reply by maliberty** [reply]
With
```
"space_to_non_fill":    3
```
I don't think it is possible to fill M1 as the rails (non-fill) look to be ~4um apart.  Is this value correct?

**Reply by gkamendje** [reply]
Thanks for the hint. Setting it to 1 make it possible to fill M1. I was wondering which LEF/Design Rule data point should drive this value.

**Reply by maliberty** [reply]
Its not in LEF which is why we need a separate config file.  Usually the DRM will have some guidance.

**Comment by gkamendje** [comment]
I noticed that some designs use the following construct in the `config.mk` file. What is the rationale behind it? 
```
ifneq ($(USE_FILL),)
export DESIGN_TYPE = CELL
else
export DESIGN_TYPE = CELL_NODEN
endif
```

**Reply by maliberty** [reply]
Those are required by a specific private PDK and aren't generally required.
